### PR TITLE
docs-infra: ngmodule pages

### PR DESCRIPTION
**Pull request by petebacondarwin**

### CODE REVIEW

**Comment by petebacondarwin** [comment]
Some examples:

* https://pr25734-9b00818.ngbuilds.io/api/common/CommonModule

screen shot 2018-08-30 at 07 43 13


* https://pr25734-9b00818.ngbuilds.io/api/forms/FormsModule

screen shot 2018-08-30 at 07 43 33

File: aio/tools/transforms/templates/api/ngmodule.template.html [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+  {$ memberHelpers.renderProperties(doc.properties, 'instance-properties', 'instance-property', 'Properties') $}
+
+  {$ memberHelpers.renderMethodDetails(versionInfo, doc.methods, 'instance-methods', 'instance-method', 'Methods') $}
+
+  {% if doc.ngmoduleOptions.providers %}
+  {$ renderTable(doc.ngmoduleOptions.providers, 'providers', 'Providers', 'Provider') $}
+  {% endif %}
+
+  {% if doc.ngmoduleOptions.imports %}
```
Comment (IgorMinar) [inline aio/tools/transforms/templates/api/ngmodule.template.html]: imports is an implementation detail of the module and should not be part of the public api docs. please remove.

File: aio/tools/transforms/templates/api/ngmodule.template.html [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+  {% if doc.ngmoduleOptions.exports %}
+  {$ renderTable(doc.ngmoduleOptions.exports, 'exports', 'Exports', 'Export') $}
+  {% endif %}
+
+  {% if doc.ngmoduleOptions.declarations %}
+  {$ renderTable(doc.ngmoduleOptions.declarations, 'declarations', 'Declarations', 'Declaration') $}
+  {% endif %}
+
+  {% if doc.ngmoduleOptions.entryComponents %}
```
Comment (IgorMinar) [inline aio/tools/transforms/templates/api/ngmodule.template.html]: private stuff, please remove

File: aio/tools/transforms/templates/api/ngmodule.template.html [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+  {% if doc.ngmoduleOptions.imports %}
+  {$ renderTable(doc.ngmoduleOptions.imports, 'imports', 'Imports', 'Import') $}
+  {% endif %}
+
+  {% if doc.ngmoduleOptions.exports %}
+  {$ renderTable(doc.ngmoduleOptions.exports, 'exports', 'Exports', 'Export') $}
+  {% endif %}
+
+  {% if doc.ngmoduleOptions.declarations %}
```
Comment (IgorMinar) [inline aio/tools/transforms/templates/api/ngmodule.template.html]: private implementation detail, please remove

**Review by IgorMinar — COMMENTED** [review]
I've thought a bit about "exports" because that is what represent the public api surface of the contents of an NgModule in respect to the user of the module.

The challenge is that in most cases, we export stuff indirectly via arrays and providers and many of these are private and shouldn't be listed in the public documentation. In the ideal case, we'd want to document that a module exports directives "A", "B" and "C", provides a provider for injection tokens "X", "Y", and "Z", and reexports another public NgModule, but this information is not trivial to extract without employing the logic from `ngc` and even then, we have internal NgModules that are not surfaced to the public api so even ngc doesn't have all the info to make this list.

I suggest that we don't even try to do that and simply just manually document these exports as three lists of symbols (directives, providers, reexported public NgModules) that then gets linked to the respective API pages and it will be up to humans to keep this list up to date. We don't changes this stuff very often so it should be ok to maintain these lists by hand.

**Comment by petebacondarwin** [comment]
So @IgorMinar are you saying that I should also remove the `Exports` and `Providers` sections, and leave the author to manually describe these in the description/usage notes?

**Review by odykyi — APPROVED** [review]

**Comment by jenniferfell** [comment]
One minor thing:  Can we choose another color for "NGMODULE" badge? We'd like to use red for re-instating the "deprecated" badge and make that more visible red (warning/caution) instead of its v5 grey.

**Comment by jenniferfell** [comment]
notes from discussion 9/6...directives: new tagging like forms library so we can get to the directives and display them. "exported from this ngmodule". no tag = doesn't appear = internal directive. tag directives and pipes now. @ngmodule on every directive or pipe, if missing then error, linked ngmodule must be public or error, then surface on directives and pipes API pages. land this PR, and then new PR for the infra for tagging, followed by PR for actual tags added.

**Comment by jenniferfell** [comment]
notes from discussion 9/6: providers: need more discussion. separate work.

**Comment by petebacondarwin** [comment]
> One minor thing: Can we choose another color for "NGMODULE" badge? We'd like to use red for re-instating the "deprecated" badge and make that more visible red (warning/caution) instead of its v5 grey.

Happy to change the colour. Please choose one and let me know :-)

**Comment by petebacondarwin** [comment]
I have added two commits. The first adds the directive/pipe export rendering that we discussed. The second fixes up the last of the directives/pipes that were missing `@ngModule` tags - it turns out most already had them!

**Comment by petebacondarwin** [comment]
Package pages now list NgModules too: https://pr25734-988458b.ngbuilds.io/api/forms

screen shot 2018-09-08 at 21 48 44

**Comment by petebacondarwin** [comment]
NgModules pages now list directives and pipes: https://pr25734-988458b.ngbuilds.io/api/common/CommonModule

screen shot 2018-09-08 at 21 51 31

screen shot 2018-09-08 at 21 51 40

**Comment by petebacondarwin** [comment]
Directives/Pipes pages list what NgModules they are exported from: https://pr25734-988458b.ngbuilds.io/api/common/NgClass

screen shot 2018-09-08 at 21 54 20

In the case of shared directives that are exported from an internal NgModule that is then re-exported to more than one public NgModule, they appear as exports on both NgModule pages and list both modules on their own page: https://pr25734-988458b.ngbuilds.io/api/forms/RequiredValidator

screen shot 2018-09-08 at 21 55 39

**Comment by IgorMinar** [comment]
All these changes look good and are way better than what I originally proposed. We could do a bit more of visual tweaking but it's fine if that's a follow up PR.

In case of public NgModules that are reexported via another public NgModule, it would be nice to list the reexported module(s). For example https://pr25734-f25f83a.ngbuilds.io/api/platform-browser/BrowserModule reexports CommonModule.

I suggest that we get this PR green and land it and then iterate on top of it. Thanks Pete!

**Comment by petebacondarwin** [comment]
@IgorMinar can you deal with the G3 presubmit requirement?

**Comment by petebacondarwin** [comment]
Also it needs your approval.

**Review by IgorMinar — APPROVED** [review]

**Pull request closed without merging**